### ISSUE TITLE
ARRAY Reference

### ISSUE THREAD

**Issue by mlam0214** [d4t4king/lynis-report-converter#1]
Dear all,

I tried to use your tool to convert Lynis report to pdf format but when I try, I got the following error:
**Can't use an undefined value as an ARRAY reference at ./lynis-report-converter.pl line 181.**

Can you help me?

Another error was present at line 86, ')' has to be deleted.

Kind regards,

**Comment by d4t4king** [comment]
@mlam0214 

Are you skipping any tests using the lynis custom profile? (e.g. ```/etc/lynis/custom.prf```)

**Comment by d4t4king** [comment]
> Another error was present at line 86, ')' has to be deleted.

This has been fixed.  Thanks for helping to catch this silly typo!

> I tried to use your tool to convert Lynis report to pdf format but when I try, I got the following error:
> Can't use an undefined value as an ARRAY reference at ./lynis-report-converter.pl line 181.

I can't directly reproduce this issue, but I believe I may have fixed it anyway.  Try a fresh pull, and let me know if you still see the issue.  If so, I will probably need to get more information from you.

**Comment by mlam0214** [comment]
Thanks for your quick answer.

The issue is still present with the new version but in another place.
Here are the errors I got:

Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric lt (<) at ./lynis-report-converter.pl line 1170.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric gt (>) at ./lynis-report-converter.pl line 1170.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric le (<=) at ./lynis-report-converter.pl line 1174.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric gt (>) at ./lynis-report-converter.pl line 1174.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric le (<=) at ./lynis-report-converter.pl line 1178.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric gt (>) at ./lynis-report-converter.pl line 1178.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric le (<=) at ./lynis-report-converter.pl line 1182.
Use of uninitialized value $lynis_report_data{"hardening_index"} in concatenation (.) or string at ./lynis-report-converter.pl line 1184.
Can't use an undefined value as an ARRAY reference at ./lynis-report-converter.pl line 1244.

**Comment by d4t4king** [comment]
Technically speaking, the "Use of uninitialized value...." aren't errors, they're warnings.  In fact, if you comment out the ```use warnings;``` line at the very beginning of the script, those issues will simply disappear.

You never answered my question from before....  Are you using a custom profile in ```/etc/lynis/custom.prf```?  What OS are you attempting to run this on?

**Comment by mlam0214** [comment]
Hi,

Thanks for the answer.

No i don't use a custom profile like /etc/lynis/custom.prf.
I'm trying to run Lynis on Latest version of Debian.

XML export works but PDF not. When I try the following command:
_./lynis-report-converter.pl -o test.pdf -p_

Here is the output:
### Looks like the audit has been run.
Outputting report to test.pdf, in PDF format.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric lt (<) at ./lynis-report-converter.pl line 1170.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric gt (>) at ./lynis-report-converter.pl line 1170.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric le (<=) at ./lynis-report-converter.pl line 1174.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric gt (>) at ./lynis-report-converter.pl line 1174.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric le (<=) at ./lynis-report-converter.pl line 1178.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric gt (>) at ./lynis-report-converter.pl line 1178.
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric le (<=) at ./lynis-report-converter.pl line 1182.
Use of uninitialized value $lynis_report_data{"hardening_index"} in concatenation (.) or string at ./lynis-report-converter.pl line 1184.
Can't use an undefined value as an ARRAY reference at ./lynis-report-converter.pl line 1244.

Thanks

**Comment by d4t4king** [comment]
OK, I use Debian _A_ _LOT_ -- it's possible even that this script was largely developed on a Debian system (or VM), so I'm a little baffled as to why I'm having difficulty reproducing your problem.

I did just commit some updates today, so if you're cloning from here, please do a ```git pull```, run the report,  and post your errors again.  "Line 1244" doesn't point to anything that could indicate an actual problem after the updates I made recently.

(Also, as previously mentioned, the "use of uninitialized....." errors are not in fact errors, they are warnings, and I've disabled warnings in my most recent commit.   You are however seeing some fatal errors and I'm struggling to understand why.  To that end, I've just downloaded the Debian 8.7.1 ISOs and will endeavor to figure it out.

**Comment by mlam0214** [comment]
I tried to export the report again and had the following errors:

Use of uninitialized value in concatenation (.) or string at ./lynis-report-conv                                                                                                                                                             erter.pl line 1349.
Can't use string ("&nbsp;") as an ARRAY ref while "strict refs" in use at ./lyni                                                                                                                                                             s-report-converter.pl line 1367.

Still also have these errors but I supposed it's because I don't use the variables?
Use of uninitialized value $lynis_report_data{"hardening_index"} in numeric lt (                                                                                                                                                             <) at ./lynis-report-converter.pl line 1175.

Thanks for you help

**Comment by d4t4king** [comment]
OK, I've turned off all the warnings now.  Or, at least I think I have.  Do another ```git pull``` and let me know what you see.

> Still also have these errors but I supposed it's because I don't use the variables?

I'm not sure what you mean by this.  If you're referring to the ```/etc/lynis/custom.prf``` I mentioned in a previous post, that shouldn't matter.  If that's not it, then I just don't know.

I just downloaded and installed Debian 9 (Stretch) and only installed the prerequisites for this script.  I was still unable to reproduce your problem.  So, I'm at a loss.  I've attempted to anticipate your issue, even though I can't reproduce it, but if you still encounter the issue (or similar) I'm not sure what else I can do.

**Comment by d4t4king** [comment]
@mlam0214 , 

I have made several updates, including situations where certain fields weren't populated in the lynis report file (```/var/log/lynis-report.dat```).

I also ran into a situation where the lynis report data file was oddly (and mercilessly) truncated.  Basically, only the lynis version, number of tests run, and a few other metadata points were saved.  Almost none of the audit data was saved to the data file.  I do not know why this happened, and if I can reproduce it again, I will be filing a bug with the lynis tool.

So, make sure you have the most recent version of the lynis tool (as of this writing, 2.4.8).  Run the lynis audit on your system again, and verify that the data file is greater than, say 100 lines:
```
$ sudo wc -l /var/log/lynis-report.dat
1135 /var/log/lynis-report.dat
$
```
(That first number in the second line should be more than 100.)
Then run my report converter script again.

Let me know how it goes.  If I don't hear from you soon, I will consider this issue resolved.

Thanks!
